jni/src/global_jni_env.h

```diff
@@ -0,0 +1,206 @@
+    jbyte (*CallNonvirtualByteMethod)(jobject arg1, jclass arg2, jmethodID arg3);
+    jbyte (*CallNonvirtualByteMethodA)(jobject obj, jclass clazz, jmethodID methodID, const jvalue * args);
+    jchar (*CallNonvirtualCharMethod)(jobject arg1, jclass arg2, jmethodID arg3);
+    jchar (*CallNonvirtualCharMethodA)(jobject obj, jclass clazz, jmethodID methodID, const jvalue * args);
+    jshort (*CallNonvirtualShortMethod)(jobject arg1, jclass arg2, jmethodID arg3);
+    jshort (*CallNonvirtualShortMethodA)(jobject obj, jclass clazz, jmethodID methodID, const jvalue * args);
+    jint (*CallNonvirtualIntMethod)(jobject arg1, jclass arg2, jmethodID arg3);
+    jint (*CallNonvirtualIntMethodA)(jobject obj, jclass clazz, jmethodID methodID, const jvalue * args);
+    jlong (*CallNonvirtualLongMethod)(jobject arg1, jclass arg2, jmethodID arg3);
+    jlong (*CallNonvirtualLongMethodA)(jobject obj, jclass clazz, jmethodID methodID, const jvalue * args);
+    jfloat (*CallNonvirtualFloatMethod)(jobject arg1, jclass arg2, jmethodID arg3);
+    jfloat (*CallNonvirtualFloatMethodA)(jobject obj, jclass clazz, jmethodID methodID, const jvalue * args);
+    jdouble (*CallNonvirtualDoubleMethod)(jobject arg1, jclass arg2, jmethodID arg3);
+    jdouble (*CallNonvirtualDoubleMethodA)(jobject obj, jclass clazz, jmethodID methodID, const jvalue * args);
+    void (*CallNonvirtualVoidMethod)(jobject arg1, jclass arg2, jmethodID arg3);
+    void (*CallNonvirtualVoidMethodA)(jobject obj, jclass clazz, jmethodID methodID, const jvalue * args);
+    jfieldID (*GetFieldID)(jclass clazz, const char * name, const char * sig);
+    jobject (*GetObjectField)(jobject obj, jfieldID fieldID);
+    jboolean (*GetBooleanField)(jobject obj, jfieldID fieldID);
+    jbyte (*GetByteField)(jobject obj, jfieldID fieldID);
+    jchar (*GetCharField)(jobject obj, jfieldID fieldID);
+    jshort (*GetShortField)(jobject obj, jfieldID fieldID);
+    jint (*GetIntField)(jobject obj, jfieldID fieldID);
+    jlong (*GetLongField)(jobject obj, jfieldID fieldID);
+    jfloat (*GetFloatField)(jobject obj, jfieldID fieldID);
+    jdouble (*GetDoubleField)(jobject obj, jfieldID fieldID);
+    void (*SetObjectField)(jobject obj, jfieldID fieldID, jobject val);
+    void (*SetBooleanField)(jobject obj, jfieldID fieldID, jboolean val);
+    void (*SetByteField)(jobject obj, jfieldID fieldID, jbyte val);
+    void (*SetCharField)(jobject obj, jfieldID fieldID, jchar val);
+    void (*SetShortField)(jobject obj, jfieldID fieldID, jshort val);
+    void (*SetIntField)(jobject obj, jfieldID fieldID, jint val);
+    void (*SetLongField)(jobject obj, jfieldID fieldID, jlong val);
+    void (*SetFloatField)(jobject obj, jfieldID fieldID, jfloat val);
+    void (*SetDoubleField)(jobject obj, jfieldID fieldID, jdouble val);
+    jmethodID (*GetStaticMethodID)(jclass clazz, const char * name, const char * sig);
```

Comment:
There's actually one for each method in JNI, and even JNI headers don't document these functions. So the user using these is supposed to know the C equivalent IG.

Reply:
Hm, we should at least refer to the official documentation then somewhere.

And we should consider documenting what we reverse engineered ourselves as about who owns the arguments and what lifetime they have. Otherwise every time we need to reverse engineer again (e.g. I'm going to forget in 3 months.) and any potential users of our generated Dart file is going to ask the same questions.